zeekpkg/manager.py

```diff
@@ -1053,6 +1102,175 @@ def load(self, pkg_path):
         LOG.debug('loaded "%s"', pkg_path)
         return ''
 
+    def _load(self, pkg_path):
+        """Mark dependent (but previously installed) packages as being "loaded".
+
+        Args:
+            pkg_path (str): the full git URL of a package or the shortened
+                path/name that refers to it within a package source.  E.g. for
+                a package source called "zeek" with package named "foo" in
+                :file:`alice/zkg.index`, the following inputs may refer
+                to the package: "foo", "alice/foo", or "zeek/alice/foo".
+
+        Returns:
+            str: empty string if the package is successfully marked as loaded,
+            else an explanation of why it failed.
+
+        """
+
+        # skip loading a package if it is not installed.
+        if not self.find_installed_package(pkg_path):
+            return 'loading dependency failed. package not installed.'
+
+        if self.load(pkg_path):
+            return 'loading dependency failed.'
+
+        for _pkg_path, _ in self.find_package_dependencies(pkg_path):
+            return self._load(_pkg_path)
+
+    def saveState(self):
+        """Save loaded state for all installed packages.
+
+        Returns:
+            dict: dictionary of load status for installed packages
+        """
+
+        saved_state = {}
+        for _pkg_name in self.pkg_dependencies:
+            ipkg = self.find_installed_package(_pkg_name)
+            if ipkg and ipkg.status:
+                saved_state[_pkg_name] = ipkg.status.is_loaded
+        return saved_state
+
+    def get_changed_state(self, saved_state, pkg_lst):
+        """Returns the list of packages that have changed loaded state.
+
+        Args:
+            saved_state (dict): dictionary of saved load state for installed
+            packages.
+
+            pkg_lst (list): list of package names to be skipped
+
+        Returns:
+            dep_listing (str): string installed packages that have changed state
+        
+        """
+        _lst = [name_from_path(_pkg_path) for _pkg_path in pkg_lst]
+        dep_listing = ''
+        for _pkg_name in self.pkg_dependencies:
+            if _pkg_name in _lst:
+                continue
+            _ipkg = self.find_installed_package(_pkg_name)
+            if _ipkg and _ipkg.status.is_loaded != saved_state[_pkg_name]:
+                dep_listing += '  {}\n'.format(_pkg_name)
+        return dep_listing               
+
+    def restoreState(self, saved_state):
```

Comment:
The `saveState` ,  `_load` (or batched load) and `restoreState` operations only operate on the installed packages. In other words, the initial install for a package is responsible for pulling down all the dependencies and thus it modifies the filesystem state. `_load` on the other hand only works with previously installed packages.

Reply:
The `_load()` itself does also modify some filesystem state.  It's not as many changes as `install()` does, but there's still some, e.g. see `_write_autoloader()` in `load()`.  That would be a filesystem change where a `saveState()`/`restoreState()` may not really restore things to a consistent state if there's a "partial/midway failure" during the batch.

Reply:
I find that both `load` and `unload` invoke the following 3 functions:

Reply:
Yes, likely can invoke those `_write` methods again to try to rollback after a failed batch, but then it's also possible for those to `_write` methods to themselves fail during the rollback attempt, so that's why I thought it may actually be easier/less-complex to avoid these types of "batched operations" on the `Manager` side of things. Could instead have the `zkg` side iterate through a "batch", emitting result at every step and also exit the program entirely if any fail.

Do agree with the general idea of implementing all operations as "atomic transactions", but may be worth treating as out of scope of this PR's own goals as a task to try later/separately since I expect there's substantial changes required to do that thoroughly and correctly everywhere.  (But also don't think it's too high priority a feature, since I can't recall any previous reported issues where it would have helped).

Reply:
I've added the `_write` methods for now. Will get rid of them once I have feedback for the rest of the PR.

Reply:
We can table this part of the conversation and I'll see where it ends up after latest round of feedback, but assuming I'll be fine with the save/restore attempts as you've got them for now.